File: pkg/client/patch.go
package client

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
)

// PatchPlan mirrors the server's response for a PATCH request.
type PatchPlan struct {
	UploadID    string            `json:"upload_id"`
	PartSize    int64             `json:"part_size"`
	UploadParts []*PatchPartURL   `json:"upload_parts"`
	CopiedParts []int             `json:"copied_parts"`
}

// PatchPartURL describes one dirty part the client must upload.
type PatchPartURL struct {
	Number    int               `json:"number"`
	URL       string            `json:"url"`
	Size      int64             `json:"size"`
	Headers   map[string]string `json:"headers,omitempty"`
	ExpiresAt string            `json:"expires_at"`
	ReadURL   string            `json:"read_url,omitempty"`
}

// PatchFile performs a partial update of a large file using S3 UploadPartCopy
// for unchanged parts. Only the dirty parts are transferred over the network.
//
// Parameters:
//   - path: remote file path
//   - newSize: total size of the file after patching
//   - dirtyParts: 1-based part numbers that have been modified
//   - readPart: callback that returns the complete data for a given part number.
//     The callback receives the part number (1-based), the part's expected size,
//     and the original data for that part (downloaded from ReadURL; nil if the
//     part is entirely new beyond the original file). It must return the final
//     part data to upload.
func (c *Client) PatchFile(ctx context.Context, path string, newSize int64, dirtyParts []int, readPart func(partNumber int, partSize int64, origData []byte) ([]byte, error), progress ProgressFunc) error {
	// Step 1: Request patch plan from server
	reqBody, err := json.Marshal(map[string]any{
		"new_size":    newSize,
		"dirty_parts": dirtyParts,
	})
	if err != nil {
		return fmt.Errorf("marshal patch request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPatch, c.url(path), bytes.NewReader(reqBody))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusAccepted {
		return readError(resp)
	}

	var plan PatchPlan
	if err := json.NewDecoder(resp.Body).Decode(&plan); err != nil {
		return fmt.Errorf("decode patch plan: %w", err)
	}

	// Step 2: Upload dirty parts concurrently
	const maxConcurrency = 4
	sem := make(chan struct{}, maxConcurrency)
	errCh := make(chan error, 1)
	var wg sync.WaitGroup

	totalParts := len(plan.UploadParts) + len(plan.CopiedParts)

	for _, part := range plan.UploadParts {
		select {
		case sem <- struct{}{}:
		case <-ctx.Done():
			wg.Wait()
			return ctx.Err()
		}

		select {
		case err := <-errCh:
			wg.Wait()
			return err
		default:
		}

		wg.Add(1)
		go func(p *PatchPartURL) {
			defer wg.Done()
			defer func() { <-sem }()

			if err := c.uploadPatchPart(ctx, p, readPart); err != nil {
				select {
				case errCh <- fmt.Errorf("part %d: %w", p.Number, err):
				default:
				}
				return
			}

			if progress != nil {
				progress(p.Number, totalParts, p.Size)
			}
		}(part)
	}

	wg.Wait()

	select {
	case err := <-errCh:
		return err
	default:
	}

	// Step 3: Complete upload (same endpoint as normal multipart)
	return c.completeUpload(ctx, plan.UploadID)
}

// uploadPatchPart downloads original data (if available), calls the readPart
// callback to get the final data, and uploads it to the presigned URL.
func (c *Client) uploadPatchPart(ctx context.Context, part *PatchPartURL, readPart func(int, int64, []byte) ([]byte, error)) error {
	// Download original part data if a read URL is provided
	var origData []byte
	if part.ReadURL != "" {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, part.ReadURL, nil)
		if err != nil {
			return fmt.Errorf("create read request: %w", err)
		}
		resp, err := c.httpClient.Do(req) // Direct to S3, no auth
		if err != nil {
			return fmt.Errorf("download original part: %w", err)
		}
		defer resp.Body.Close()
		if resp.StatusCode >= 300 {
			return fmt.Errorf("download original part: HTTP %d", resp.StatusCode)
		}
		origData, err = io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("read original part body: %w", err)
		}
	}

	// Get the final part data from the callback
	data, err := readPart(part.Number, part.Size, origData)
	if err != nil {
		return fmt.Errorf("readPart callback: %w", err)
	}

	// Upload to presigned URL
	h := sha256.Sum256(data)
	checksum := base64.StdEncoding.EncodeToString(h[:])

	req, err := http.NewRequestWithContext(ctx, http.MethodPut, part.URL, bytes.NewReader(data))
	if err != nil {
		return err
	}
	for k, v := range part.Headers {
		if strings.EqualFold(k, "host") {
			continue
		}
		req.Header.Set(k, v)
	}
	req.ContentLength = int64(len(data))
	req.Header.Set("x-amz-checksum-sha256", checksum)

	resp, err := c.httpClient.Do(req) // Direct to S3, no auth
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("upload part: HTTP %d: %s", resp.StatusCode, string(body))
	}

	return nil
}

File: clients/drive9-py/drive9/patch.py
"""Patch upload helpers for partial file updates."""

import base64
import hashlib
import json
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Optional, Callable

from .exceptions import StatusError
from .models import PatchPartURL

ProgressFunc = Callable[[int, int, int], None]
_ReadPartFunc = Callable[[int, int, Optional[bytes]], bytes]


class PatchMixin:
    """Mixin that adds patch file methods to Client."""

    def patch_file(
        self,
        path: str,
        new_size: int,
        dirty_parts: list,
        read_part: _ReadPartFunc,
        progress: Optional[ProgressFunc] = None,
        part_size: int = 0,
    ) -> None:
        """Perform a partial update of a large file.

        Args:
            path: Remote file path.
            new_size: Total size of the file after patching.
            dirty_parts: 1-based part numbers that have been modified.
            read_part: Callback that returns the complete data for a given part.
                Receives (part_number, part_size, orig_data).
            progress: Optional progress callback.
            part_size: Optional part size hint.
        """
        patch_req = {
            "new_size": new_size,
            "dirty_parts": dirty_parts,
        }
        if part_size > 0:
            patch_req["part_size"] = part_size

        resp = self._request(
            "PATCH",
            self._url(path),
            data=json.dumps(patch_req),
            headers={"Content-Type": "application/json"},
        )
        if resp.status_code != 202:
            raise StatusError(
                resp.text,
                status_code=resp.status_code,
                response=resp,
            )
        plan = resp.json()

        upload_parts = [
            PatchPartURL(
                number=p["number"],
                url=p["url"],
                size=p["size"],
                headers=p.get("headers"),
                expires_at=p.get("expires_at"),
                read_url=p.get("read_url"),
                read_headers=p.get("read_headers"),
            )
            for p in plan.get("upload_parts", [])
        ]
        copied_parts = plan.get("copied_parts", [])
        total_parts = len(upload_parts) + len(copied_parts)

        max_concurrency = 4
        err = None
        err_lock = threading.Lock()

        def upload_one(part: PatchPartURL):
            nonlocal err
            if err is not None:
                return
            try:
                self._upload_patch_part(part, read_part)
                if progress:
                    progress(part.number, total_parts, part.size)
            except Exception as exc:
                with err_lock:
                    if err is None:
                        err = exc

        executor = ThreadPoolExecutor(max_workers=max_concurrency)
        futures = {executor.submit(upload_one, p): p for p in upload_parts}
        try:
            for future in as_completed(futures):
                future.result()
        except Exception:
            executor.shutdown(wait=False)
            raise
        executor.shutdown(wait=True)

        if err is not None:
            raise err

        self._complete_upload(plan["upload_id"])

    def _upload_patch_part(self, part: PatchPartURL, read_part: _ReadPartFunc) -> None:
        orig_data = None
        if part.read_url:
            headers = {}
            if part.read_headers:
                for k, v in part.read_headers.items():
                    if k.lower() == "host":
                        continue
                    headers[k] = v
            resp = self.session.get(part.read_url, headers=headers)
            if resp.status_code >= 300:
                raise StatusError(
                    f"download original part: HTTP {resp.status_code}: {resp.text}",
                    status_code=resp.status_code,
                    response=resp,
                )
            orig_data = resp.content

        data = read_part(part.number, part.size, orig_data)
        checksum = base64.b64encode(hashlib.sha256(data).digest()).decode("ascii")

        headers = {"x-amz-checksum-sha256": checksum}
        if part.headers:
            for k, v in part.headers.items():
                if k.lower() == "host":
                    continue
                headers[k] = v

        resp = self.session.put(part.url, data=data, headers=headers)
        if resp.status_code >= 300:
            raise StatusError(
                f"upload part: HTTP {resp.status_code}: {resp.text}",
                status_code=resp.status_code,
                response=resp,
            )
